x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/CoordinatorProxyAction.java

```diff
@@ -0,0 +1,154 @@
+
+        private final Coordinator coordinator;
+
+        @Inject
+        public TransportAction(TransportService transportService, ActionFilters actionFilters, Coordinator coordinator) {
+            super(NAME, transportService, actionFilters, (Writeable.Reader<SearchRequest>) SearchRequest::new);
+            this.coordinator = coordinator;
+        }
+
+        @Override
+        protected void doExecute(Task task, SearchRequest request, ActionListener<SearchResponse> listener) {
+            assert Thread.currentThread().getName().contains(ThreadPool.Names.WRITE);
+            coordinator.schedule(request, listener);
+        }
+    }
+
+    public static class Coordinator {
+
+        final Client client;
+        final int maxLookupsPerRequest;
+        final int maxNumberOfConcurrentRequests;
+        final BlockingQueue<Item> queue;
+        final AtomicInteger numberOfOutstandingRequests = new AtomicInteger(0);
+
+        public Coordinator(Client client, Settings settings) {
+            this(client,
+                EnrichPlugin.COORDINATOR_PROXY_MAX_LOOKUPS_PER_REQUEST.get(settings),
+                EnrichPlugin.COORDINATOR_PROXY_MAX_CONCURRENT_REQUESTS.get(settings),
+                EnrichPlugin.COORDINATOR_PROXY_QUEUE_CAPACITY.get(settings));
+        }
+
+        Coordinator(Client client, int maxLookupsPerRequest, int maxNumberOfConcurrentRequests, int queueCapacity) {
+            this.client = client;
+            this.maxLookupsPerRequest = maxLookupsPerRequest;
+            this.maxNumberOfConcurrentRequests = maxNumberOfConcurrentRequests;
+            this.queue = new LinkedBlockingQueue<>(queueCapacity);
+        }
+
+        void schedule(SearchRequest searchRequest, ActionListener<SearchResponse> listener) {
+            // Use put(...), because if queue is full then this method will wait until a free slot becomes available
+            // The calling thread here is a write thread (write tp is used by ingest) and
+            // this will great natural back pressure from the enrich processor.
+            // If there are no write threads available then write requests with ingestion will fail with 429 error code.
+            try {
+                queue.put(new Item(searchRequest, listener));
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException("unable to add item to queue", e);
+            }
+            coordinateLookups();
+        }
+
+        synchronized void coordinateLookups() {
+            while (queue.isEmpty() == false &&
+                numberOfOutstandingRequests.get() < maxNumberOfConcurrentRequests) {
+
+                final List<Item> items = new ArrayList<>();
+                queue.drainTo(items, maxLookupsPerRequest);
+                final MultiSearchRequest multiSearchRequest = new MultiSearchRequest();
+                items.forEach(item -> multiSearchRequest.add(item.searchRequest));
```

Comment:
Do you think it'd be worth trying to deduplicate search requests here? Or would it be handled by the cache you plan to add on top anyway?

Reply:
Yes, I think it is worth to deduplicate search requests here. I will address this in a follow-up change. 

> Or would it be handled by the cache you plan to add on top anyway?

The cache would help to, but I'm hoping that this change and follow up changes will make the enrich processor depend less on a cache for performance.